crates/e2e/src/backend_calls.rs

```diff
@@ -0,0 +1,320 @@
+            self.message,
+            self.value,
+            self.storage_deposit_limit,
+        )
+        .await;
+        let gas_required: Weight =
+            self.extra_gas_portion
+                .map_or(dry_run.exec_result.gas_required, |margin| {
+                    let gas = dry_run.exec_result.gas_required;
+                    gas + (gas / 100 * margin)
+                });
+        let call_result = C::bare_call(
+            self.client,
+            self.caller,
+            self.message,
+            self.value,
+            gas_required,
+            self.storage_deposit_limit,
+        )
+        .await?;
+        Ok(CallResult {
+            dry_run,
+            events: call_result,
+        })
+    }
+
+    /// Dry run the call.
+    pub async fn dry_run(&mut self) -> CallDryRunResult<E, RetType>
+    where
+        CallBuilderFinal<E, Args, RetType>: Clone,
+    {
+        C::bare_call_dry_run(
+            self.client,
+            self.caller,
+            self.message,
+            self.value,
+            self.storage_deposit_limit,
+        )
+        .await
+    }
+}
+
+/// Allows to build an end-to-end instantiation call using a builder pattern.
+pub struct InstantiateBuilder<'a, E, Contract, Args, R, C>
+where
+    E: Environment,
+    Args: Encode + Clone,
+
+    C: ContractsBackend<E>,
+{
+    client: &'a mut C,
+    caller: &'a Keypair,
+    contract_name: &'a str,
+    constructor: CreateBuilderPartial<E, Contract, Args, R>,
+    value: E::Balance,
+    extra_gas_portion: Option<u64>,
+    storage_deposit_limit: Option<E::Balance>,
+}
+
+impl<'a, E, Contract, Args, R, C> InstantiateBuilder<'a, E, Contract, Args, R, C>
+where
+    E: Environment,
+    Args: Encode + Clone + Send + Sync,
+
+    C: ContractsBackend<E>,
+{
+    /// Initialize a call builder with essential values.
+    pub fn new(
+        client: &'a mut C,
+        caller: &'a Keypair,
+        contract_name: &'a str,
+        constructor: CreateBuilderPartial<E, Contract, Args, R>,
+    ) -> InstantiateBuilder<'a, E, Contract, Args, R, C>
+    where
+        E::Balance: From<u32>,
+    {
+        Self {
+            client,
+            caller,
+            contract_name,
+            constructor,
+            value: 0u32.into(),
+            extra_gas_portion: None,
+            storage_deposit_limit: None,
+        }
+    }
+
+    /// Provide value with a call
+    pub fn value(&mut self, value: E::Balance) -> &mut Self {
+        self.value = value;
+        self
+    }
+
+    /// Increases the gas limit marginally by a specified percent.
+    /// Useful when the message's gas usage depends on the runtime state
+    /// and the dry run does not produce an accurate gas estimate.
+    ///
+    /// # Example
+    ///
+    /// With dry run gas estimate of `100` units and `5`% extra gas portion specified,
+    /// the set gas limit becomes `105` units
+    pub fn extra_gas_portion(&mut self, per_cent: u64) -> &mut Self {
```

Comment:
Is it possible to predict the extra gas portion based on the block saturation level?

Reply:
If by block saturation you mean how much block weight is consumed for a given block, then I don't think it is related to the gas estimate. The reason why we want to add extra gas margin is because dry-run does not produce an accurate estimate. This can happend when the state changes when we submit an on-chain call and the contract performs some extra computation consuming extra gas

Reply:
I thought that this extra gas consumption could be related to https://spec.polkadot.network/id-weights#id-fee-multiplier. I encountered issues while conducting contract benchmarks on Moonbeam, where the same call on the same contract could consume varying amounts of gas due to network load.

Reply:
This is part of polkadot runtime, not the contract gas consumption which is completely different thing